### PR TITLE
feat: custom replay masking rules

### PR DESCRIPTION
**Pull request by vaind**
## :scroll: Description

- **refactor: change redaction logic to custom filters**
- **refactor widget filter to handle errors gracefully**
- **add an option to disable masking asset images**
- **add new masking config class**
- **expose masking configs to users**

## :bulb: Motivation and Context

part of https://github.com/getsentry/sentry-dart/issues/1193#issuecomment-2332229529

## :green_heart: How did you test it?

manually & unit tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [x] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps

I didn't want to do more than necessary in this PR but I think we should move the masking logic & widget filter out of the replay directory & move the configuration out of the replay configuration in order to support #1956

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,24 @@
...
+      options.experimental.replay.unmask<Image>();
+      options.experimental.replay.maskCallback<Text>(
+          (Element element, Text widget) =>
+              (widget.data?.contains('secret') ?? false)
+                  ? SentryMaskingDecision.mask
+                  : SentryMaskingDecision.continueProcessing);
+    },
+    appRunner: () => runApp(MyApp()),
+  );
+  ```
```
Comment (buenaflor) [inline CHANGELOG.md:27]: can we also add an example with `SentryMask` / `SentryUnmask` widget wrapping

File: flutter/lib/src/replay/masking_config.dart [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  /// be checked for children).
+  unmask,
+
+  /// Don't make a decision - continue checking other rules and children.
+  continueProcessing
+}
+
+@internal
+abstract class SentryMaskingRule<T extends Widget> {
+  bool appliesTo(Widget widget) => widget is T;
```
Comment (buenaflor) [inline flutter/lib/src/replay/masking_config.dart:51]:
```suggestion
  @pragma('vm:prefer-inline')
  bool appliesTo(Widget widget) => widget is T;
```

would inlining here be beneficial?

File: CHANGELOG.md (hunk 1 above)
Comment (vaind) [reply CHANGELOG.md:27]: added a note, although the use of these widgets is pretty simple

Comment (buenaflor) [reply CHANGELOG.md:27]: thx 👍 , just wanted it to be clear which alternative configurations there are

**Review by buenaflor — APPROVED** [review]
👍 nice

**Comment by vaind** [comment]
> 👍 nice

thanks for the review. I've noticed one important code branch is missing in tests so I've added it.  Please have a look and feel free to merge if it looks good to you: [`be99d93` (#2324)](https://github.com/getsentry/sentry-dart/pull/2324/commits/be99d931475db82c7741e6e03f9a01d7380d207d)

**Comment by buenaflor** [comment]
additional test looks good, I'll merge this now. 

CI is failing but that's because setup-java is having problems https://github.com/actions/setup-java/issues/692

**Pull request merged**